File: output/cpp/tests/33_doc_history_basic.cpp
#include <string>
#include <unordered_map>
#include <vector>
#include <any>
#include <memory>


#include <iostream>
#include <string>
#include <cassert>

// Documentation Example: History with push$/pop$ (History201)

class HistoryBasicFrameEvent {
public:
    std::string _message;
    std::unordered_map<std::string, std::any> _parameters;

    HistoryBasicFrameEvent(const std::string& message, std::unordered_map<std::string, std::any> params = {})
        : _message(message), _parameters(std::move(params)) {}
};

class HistoryBasicFrameContext {
public:
    HistoryBasicFrameEvent _event;
    std::any _return;
    std::unordered_map<std::string, std::any> _data;

    HistoryBasicFrameContext(HistoryBasicFrameEvent event, std::any default_return = {})
        : _event(std::move(event)), _return(std::move(default_return)) {}
};

class HistoryBasicCompartment {
public:
    std::string state;
    std::unordered_map<std::string, std::any> state_args;
    std::unordered_map<std::string, std::any> state_vars;
    std::unordered_map<std::string, std::any> enter_args;
    std::unordered_map<std::string, std::any> exit_args;
    std::unique_ptr<HistoryBasicFrameEvent> forward_event;
    std::unique_ptr<HistoryBasicCompartment> parent_compartment;

    explicit HistoryBasicCompartment(const std::string& state) : state(state) {}

    std::unique_ptr<HistoryBasicCompartment> clone() const {
        auto c = std::make_unique<HistoryBasicCompartment>(state);
        c->state_args = state_args;
        c->state_vars = state_vars;
        c->enter_args = enter_args;
        c->exit_args = exit_args;
        return c;
    }
};

class HistoryBasic {
private:
    std::vector<std::unique_ptr<HistoryBasicCompartment>> _state_stack;
    std::unique_ptr<HistoryBasicCompartment> __compartment;
    std::unique_ptr<HistoryBasicCompartment> __next_compartment;
    std::vector<HistoryBasicFrameContext> _context_stack;

    void __kernel(HistoryBasicFrameEvent& __e) {
        __router(__e);
        while (__next_compartment) {
            auto next_compartment = std::move(__next_compartment);
            HistoryBasicFrameEvent exit_event("<$");
            __router(exit_event);
            __compartment = std::move(next_compartment);
            if (!__compartment->forward_event) {
                HistoryBasicFrameEvent enter_event("$>");
                __router(enter_event);
            } else {
                auto forward_event = std::move(__compartment->forward_event);
                if (forward_event->_message == "$>") {
                    __router(*forward_event);
                } else {
                    HistoryBasicFrameEvent enter_event("$>");
                    __router(enter_event);
                    __router(*forward_event);
                }
            }
        }
    }

    void __router(HistoryBasicFrameEvent& __e) {
        const auto& state_name = __compartment->state;
        if (state_name == "A") {
            _state_A(__e);
        } else if (state_name == "B") {
            _state_B(__e);
        } else if (state_name == "C") {
            _state_C(__e);
        }
    }

    void __transition(std::unique_ptr<HistoryBasicCompartment> next) {
        __next_compartment = std::move(next);
    }

    void _state_C(HistoryBasicFrameEvent& __e) {
        if (__e._message == "get_state") {
            _context_stack.back()._return = std::any(std::string("C"));
            return;;
        } else if (__e._message == "return_back") {
            auto __popped = std::move(_state_stack.back()); _state_stack.pop_back();
            __transition(std::move(__popped));
            return;
        }
    }

    void _state_A(HistoryBasicFrameEvent& __e) {
        if (__e._message == "get_state") {
            _context_stack.back()._return = std::any(std::string("A"));
            return;;
        } else if (__e._message == "gotoB") {
            auto __new_compartment = std::make_unique<HistoryBasicCompartment>("B");
            __new_compartment->parent_compartment = __compartment->clone();
            __transition(std::move(__new_compartment));
            return;
        } else if (__e._message == "gotoC_from_A") {
            _state_stack.push_back(__compartment->clone());
            auto __new_compartment = std::make_unique<HistoryBasicCompartment>("C");
            __new_compartment->parent_compartment = __compartment->clone();
            __transition(std::move(__new_compartment));
            return;
        }
    }

    void _state_B(HistoryBasicFrameEvent& __e) {
        if (__e._message == "get_state") {
            _context_stack.back()._return = std::any(std::string("B"));
            return;;
        } else if (__e._message == "gotoC_from_B") {
            _state_stack.push_back(__compartment->clone());
            auto __new_compartment = std::make_unique<HistoryBasicCompartment>("C");
            __new_compartment->parent_compartment = __compartment->clone();
            __transition(std::move(__new_compartment));
            return;
        }
    }

public:
    HistoryBasic() {
        __compartment = std::make_unique<HistoryBasicCompartment>("A");
        HistoryBasicFrameEvent __frame_event("$>");
        HistoryBasicFrameContext __ctx(std::move(__frame_event));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    void gotoC_from_A() {
        HistoryBasicFrameEvent __e("gotoC_from_A");
        HistoryBasicFrameContext __ctx(std::move(__e));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    void gotoC_from_B() {
        HistoryBasicFrameEvent __e("gotoC_from_B");
        HistoryBasicFrameContext __ctx(std::move(__e));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    void gotoB() {
        HistoryBasicFrameEvent __e("gotoB");
        HistoryBasicFrameContext __ctx(std::move(__e));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    void return_back() {
        HistoryBasicFrameEvent __e("return_back");
        HistoryBasicFrameContext __ctx(std::move(__e));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    std::string get_state() {
        HistoryBasicFrameEvent __e("get_state");
        HistoryBasicFrameContext __ctx(std::move(__e), std::any(std::string()));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        auto __result = std::any_cast<std::string>(std::move(_context_stack.back()._return));
        _context_stack.pop_back();
        return __result;
    }
};

int main() {
    printf("=== Test 33: Doc History Basic ===\n");
    HistoryBasic h;

    if (h.get_state() != "A") {
        printf("FAIL: Expected 'A', got '%s'\n", h.get_state().c_str());
        assert(false);
    }

    h.gotoC_from_A();
    if (h.get_state() != "C") {
        printf("FAIL: Expected 'C', got '%s'\n", h.get_state().c_str());
        assert(false);
    }

    h.return_back();
    if (h.get_state() != "A") {
        printf("FAIL: Expected 'A' after pop, got '%s'\n", h.get_state().c_str());
        assert(false);
    }

    h.gotoB();
    if (h.get_state() != "B") {
        printf("FAIL: Expected 'B', got '%s'\n", h.get_state().c_str());
        assert(false);
    }

    h.gotoC_from_B();
    if (h.get_state() != "C") {
        printf("FAIL: Expected 'C', got '%s'\n", h.get_state().c_str());
        assert(false);
    }

    h.return_back();
    if (h.get_state() != "B") {
        printf("FAIL: Expected 'B' after pop, got '%s'\n", h.get_state().c_str());
        assert(false);
    }

    printf("PASS: History push/pop works correctly\n");
    return 0;
}

File: output/go/tests/33_doc_history_basic.go

package main

import "fmt"
import "os"

type HistoryBasicFrameEvent struct {
    _message    string
    _parameters map[string]any
}

type HistoryBasicFrameContext struct {
    _event  HistoryBasicFrameEvent
    _return any
    _data   map[string]any
}

type HistoryBasicCompartment struct {
    state            string
    stateArgs        map[string]any
    stateVars        map[string]any
    enterArgs        map[string]any
    exitArgs         map[string]any
    forwardEvent     *HistoryBasicFrameEvent
    parentCompartment *HistoryBasicCompartment
}

func newHistoryBasicCompartment(state string) *HistoryBasicCompartment {
    return &HistoryBasicCompartment{
        state:    state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
    }
}

func (c *HistoryBasicCompartment) copy() *HistoryBasicCompartment {
    nc := &HistoryBasicCompartment{
        state: c.state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
        forwardEvent:     c.forwardEvent,
        parentCompartment: c.parentCompartment,
    }
    for k, v := range c.stateArgs { nc.stateArgs[k] = v }
    for k, v := range c.stateVars { nc.stateVars[k] = v }
    for k, v := range c.enterArgs { nc.enterArgs[k] = v }
    for k, v := range c.exitArgs { nc.exitArgs[k] = v }
    return nc
}

type HistoryBasic struct {
    _state_stack []*HistoryBasicCompartment
    __compartment *HistoryBasicCompartment
    __next_compartment *HistoryBasicCompartment
    _context_stack []HistoryBasicFrameContext
}

func NewHistoryBasic() *HistoryBasic {
    s := &HistoryBasic{}
    s._state_stack = make([]*HistoryBasicCompartment, 0)
    s._context_stack = make([]HistoryBasicFrameContext, 0)
    s.__compartment = newHistoryBasicCompartment("A")
    s.__next_compartment = nil
    __frame_event := HistoryBasicFrameEvent{_message: "$>", _parameters: nil}
    __ctx := HistoryBasicFrameContext{_event: __frame_event, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return s
}

func (s *HistoryBasic) __kernel(__e *HistoryBasicFrameEvent) {
    s.__router(__e)
    for s.__next_compartment != nil {
        next_compartment := s.__next_compartment
        s.__next_compartment = nil
        exit_event := &HistoryBasicFrameEvent{_message: "<$", _parameters: s.__compartment.exitArgs}
        s.__router(exit_event)
        s.__compartment = next_compartment
        if s.__compartment.forwardEvent == nil {
            enter_event := &HistoryBasicFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
            s.__router(enter_event)
        } else {
            forward_event := s.__compartment.forwardEvent
            s.__compartment.forwardEvent = nil
            if forward_event._message == "$>" {
                s.__router(forward_event)
            } else {
                enter_event := &HistoryBasicFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
                s.__router(enter_event)
                s.__router(forward_event)
            }
        }
    }
}

func (s *HistoryBasic) __router(__e *HistoryBasicFrameEvent) {
    switch s.__compartment.state {
    case "A":
        s._state_A(__e)
    case "B":
        s._state_B(__e)
    case "C":
        s._state_C(__e)
    }
}

func (s *HistoryBasic) __transition(next *HistoryBasicCompartment) {
    s.__next_compartment = next
}

func (s *HistoryBasic) GotoCFromA() {
    __e := HistoryBasicFrameEvent{_message: "GotoCFromA"}
    __ctx := HistoryBasicFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
}

func (s *HistoryBasic) GotoCFromB() {
    __e := HistoryBasicFrameEvent{_message: "GotoCFromB"}
    __ctx := HistoryBasicFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
}

func (s *HistoryBasic) GotoB() {
    __e := HistoryBasicFrameEvent{_message: "GotoB"}
    __ctx := HistoryBasicFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
}

func (s *HistoryBasic) ReturnBack() {
    __e := HistoryBasicFrameEvent{_message: "ReturnBack"}
    __ctx := HistoryBasicFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
}

func (s *HistoryBasic) GetState() string {
    __e := HistoryBasicFrameEvent{_message: "GetState"}
    __ctx := HistoryBasicFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result string
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(string) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *HistoryBasic) _state_A(__e *HistoryBasicFrameEvent) {
    if __e._message == "GetState" {
        s._context_stack[len(s._context_stack)-1]._return = "A"
        return
    } else if __e._message == "GotoB" {
        __compartment := newHistoryBasicCompartment("B")
        __compartment.parentCompartment = s.__compartment.copy()
        s.__transition(__compartment)
        return
    } else if __e._message == "GotoCFromA" {
        s._state_stack = append(s._state_stack, s.__compartment.copy())
        __compartment := newHistoryBasicCompartment("C")
        __compartment.parentCompartment = s.__compartment.copy()
        s.__transition(__compartment)
        return
    }
}

func (s *HistoryBasic) _state_B(__e *HistoryBasicFrameEvent) {
    if __e._message == "GetState" {
        s._context_stack[len(s._context_stack)-1]._return = "B"
        return
    } else if __e._message == "GotoCFromB" {
        s._state_stack = append(s._state_stack, s.__compartment.copy())
        __compartment := newHistoryBasicCompartment("C")
        __compartment.parentCompartment = s.__compartment.copy()
        s.__transition(__compartment)
        return
    }
}

func (s *HistoryBasic) _state_C(__e *HistoryBasicFrameEvent) {
    if __e._message == "GetState" {
        s._context_stack[len(s._context_stack)-1]._return = "C"
        return
    } else if __e._message == "ReturnBack" {
        __popped := s._state_stack[len(s._state_stack)-1]
        s._state_stack = s._state_stack[:len(s._state_stack)-1]
        s.__transition(__popped)
        return
    }
}

func main() {
	fmt.Println("=== Test 33: Doc History Basic ===")
	h := NewHistoryBasic()

	// Start in A
	if h.GetState() != "A" {
		fmt.Printf("FAIL: Expected 'A', got '%s'\n", h.GetState())
		os.Exit(1)
	}

	// Go to C from A (push A)
	h.GotoCFromA()
	if h.GetState() != "C" {
		fmt.Printf("FAIL: Expected 'C', got '%s'\n", h.GetState())
		os.Exit(1)
	}

	// Return back (pop to A)
	h.ReturnBack()
	if h.GetState() != "A" {
		fmt.Printf("FAIL: Expected 'A' after pop, got '%s'\n", h.GetState())
		os.Exit(1)
	}

	// Now go to B
	h.GotoB()
	if h.GetState() != "B" {
		fmt.Printf("FAIL: Expected 'B', got '%s'\n", h.GetState())
		os.Exit(1)
	}

	// Go to C from B (push B)
	h.GotoCFromB()
	if h.GetState() != "C" {
		fmt.Printf("FAIL: Expected 'C', got '%s'\n", h.GetState())
		os.Exit(1)
	}

	// Return back (pop to B)
	h.ReturnBack()
	if h.GetState() != "B" {
		fmt.Printf("FAIL: Expected 'B' after pop, got '%s'\n", h.GetState())
		os.Exit(1)
	}

	fmt.Println("PASS: History push/pop works correctly")
}
